demos/fee-calculator/src/App.tsx

```diff
@@ -0,0 +1,225 @@
+import { useState } from 'react';
+import './App.css';
+import Parameter from './Parameter';
+import Output from './Output';
+import githubLogoUrl from './githubLogo.svg';
+
+const defaults = {
+  ethPrice: 2200,
+  l1GasPrice: 40,
+  l2GasPrice: 0.01,
+  l2CompressionRatio: 0.7,
+  calldataCostReduction4844: 1,
+  bundleSize: 10,
+  bundlerProfitMargin: 0.05,
+};
+
+const constants = {
+  transferGas: 21_000,
+  transferEffectiveBytes: 120,
+  gasPerByte: 16,
+  l2FixedOverhead: 188, // Fixed overhead of L1 gas (before compression)
+};
+
+const App = () => {
+  const [ethPrice, setEthPrice] = useState(defaults.ethPrice);
+  const [l1GasPrice, setL1GasPrice] = useState(defaults.l1GasPrice);
+  const [l2GasPrice, setL2GasPrice] = useState(defaults.l2GasPrice);
+
+  const [l2CompressionRatio, setL2CompressionRatio] = useState(
+    defaults.l2CompressionRatio,
+  );
+
+  const [calldataCostReduction4844, setCalldataCostReduction4844] = useState(
+    defaults.calldataCostReduction4844,
+  );
+
+  const [bundleSize, setBundleSize] = useState(defaults.bundleSize);
+
+  const [bundlerProfitMargin, setBundlerProfitMargin] = useState(
+    defaults.bundlerProfitMargin,
+  );
+
+  const l1TransferFee = ethPrice * l1GasPrice * 1e-9 * constants.transferGas;
+
+  const l2TransferFee = (() => {
+    const l1Gas =
+      (l2CompressionRatio / calldataCostReduction4844) *
+      (constants.transferEffectiveBytes * constants.gasPerByte +
+        constants.l2FixedOverhead);
```

Comment:
Can you explain this calculation?

Reply:
Sure. The starting point is `constants.transferEffectiveBytes * constants.gasPerByte`. This is a crude calculation of the amount of L1 gas that the L2 needs to charge for.

Multiplying this by `l2CompressionRatio / calldataCostReduction4844` accounts for L2-level compression (gzip/brotli/etc) and the future scale change that blobs (4844) will introduce. Currently we just default `calldataCostReduction4844` to 1 so that it has no impact, because we don't have blobs yet.

Finally, there's an additional overhead that needs to get scaled in the same way. I imagine this exists because of extra metadata that needs to be posted to L1, and the L2 is amortizing this cost over each transaction equally.

This maps exactly onto Optimism's model if you use the right parameters.

I'm planning to change the model slightly to reflect this work I've been doing to [document wax compression](https://hackmd.io/lFYwpXVWS6iJ6w7HnYM0YQ?view), because it lends itself to direct measurement instead of knowing the internals of the L2:

<img width="588" alt="Screen Shot 2024-01-09 at 10 21 24" src="https://github.com/getwax/wax/assets/9291586/1683ac13-d850-4b7a-a6dd-34c618cb57db">

...

<img width="512" alt="Screen Shot 2024-01-09 at 10 23 01" src="https://github.com/getwax/wax/assets/9291586/c0173e9f-8057-4c6a-9d29-467c69de226f">

Reply:
Thanks for the explanation - makes sense now.

Only follow up question I have is if `l2CompressionRatio` accounts for the compression done by the L2, how did you arrive at the value of `0.7`? Was it an estimate given by optimism/arbitrum

Reply:
It's based on optimism since they give an explicit value on-chain and in the block explorer of 0.684. I expected arbitrum to give similar results since brotli/gzip are fairly similar. Further testing seems to show optimism giving better results here and I'm not sure why.